### ISSUE TITLE
Check possibility of having separate releases with and without database functionality

### ISSUE THREAD

**Issue by pouriyajamshidi** [pouriyajamshidi/tcping#266]
## Summary

The database functionality, `--db` flag, has added a few megabytes (especially for Windows) to the binary size and it is a feature that a very few people would use.

It would be nice to be able to offer separate versions, with and without the Database functionality.

Hopefully, this can be achieved with minimal changes to the code base and can be automated using the Makefile we have.

**Comment by pouriyajamshidi** [comment]
Hey @cyqsimon,

Would you have any knowledge on this? Or is this not a possibility in our current state?

**Comment by cyqsimon** [comment]
Yeah, so there are two parts to this:

1. use Go's preferred conditional compilation mechanism to allow us to selectively enable the feature when building on CLI.

I'm not very familiar with coding in Go, but a quick Google search seem to show it's not too difficult:

- https://pkg.go.dev/cmd/go#hdr-Build_constraints
- https://stackoverflow.com/a/67937234/5637701
- https://dev.to/jacktt/go-build-in-advance-4o8n#iii-build-tags

2. Modify the makefile to output distinct artefacts with this feature enabled/disabled

This one I can help with if you'd like to. It's a bit annoying (as is always the case with makefiles) but it's not difficult.

**Comment by pouriyajamshidi** [comment]
Thanks a lot mate!

I did not want to push this to you but figured you might have more knowledge on this than I do.

If you would like to take this on, you are more than welcome and I am very thankful for it.

**Comment by cyqsimon** [comment]
> If you would like to take this on, you are more than welcome and I am very thankful for it.

Just to be clear, I can help with the makefile changes. But you probably wouldn't want me messing with Go code.

**Comment by pouriyajamshidi** [comment]
ah gotcha!

fair

**Comment by SYSHIL** [comment]
Hey @pouriyajamshidi I can try and help here if this already hasn't been picked up

**Comment by pouriyajamshidi** [comment]
Hey @SYSHIL,

Yes, please go ahead.

We can focus on a conditional build for DB and non-DB versions for now. Our friend cyqsimon will probably help with the Makefile.

**Comment by Ilhan-Personal** [comment]
@pouriyajamshidi I was wondering if we could refactor some of our existing code before we dove into this?

Right now everything is just in one package and the main file is getting bigger and bigger, I was hoping we could restructure our code a little, Something like this? 

![image](https://github.com/user-attachments/assets/c7359a2d-7be6-4a4b-b540-486d66081040)

Also it would be nice to adopt a functional options pattern to initialize our tcping struct https://golang.cafe/blog/golang-functional-options-pattern.html
What do you think?

**Comment by pouriyajamshidi** [comment]
Yes, I 100% agree and like the functional pattern. This is bound to happen for offering `tcping` a library too. However, at this specific time, I don't want to mix these.

If this is a blocker for you, give me some time to come up with a plan. My main goal is to keep the git metadata as much as possible.

It could be good to open an issue for this.

**Comment by Ilhan-Personal** [comment]
Thanks, Hmm its not a direct blocker persay but it would be easier to implement and read if our codebase had more well-defined boundaries i think.
I don't feel strongly about restructuring the code right now, Let me try my hand at this this week and if I'm not able to and feel the solution is "hacky" ill reach out to you ?

Will also open an issue...Thanks man

**Comment by Ilhan-Personal** [comment]
So I think we can pull this off, But it feels hacky to do it without restructuring our code....what do you think?

**Comment by pouriyajamshidi** [comment]
hacky in what way? Can you share a small example?

refactoring will be done but not sure how soon

**Comment by Ilhan-Personal** [comment]
Sure, We could really just duplicate the main tcping.go file and strip the duplicated file from db related code and control compilation of separate releases through go build flags but this doesn't look elegant to me 
https://github.com/pouriyajamshidi/tcping/pull/281

**Comment by Ilhan-Personal** [comment]
Heyy @pouriyajamshidi how are you doing?
Checking in to see if there's work here that I can help with :)

**Comment by pouriyajamshidi** [comment]
Hey @Ilhan-Personal,

I am fine mate. what about you?

Well, I've been working on v3 (there's v3 branch you can check) but so far life has been busy, keeping me from continuing.
Need to gather my thoughts on how to proceed since I wanna add HTTP and UDP support.

Thank you for keeping an eye on this and willing to help.

**Comment by Ilhan-Personal** [comment]
I'm doing good too :) 
had a busy few weeks but now I'm back !

> Well, I've been working on v3 (there's v3 branch you can check) but so far life has been busy, keeping me from continuing.
> Need to gather my thoughts on how to proceed since I wanna add HTTP and UDP support.
Ohh that sounds exciting, will take a look at v3

> Thank you for keeping an eye on this and willing to help.
Of course, Your project has been super helpful for me in the past.
Please do let me know if there's anything I can do or even discuss stratergies over a call
